File: backend/src/main/java/com/example/demo/controller/AuthController.java
package com.example.demo.controller;

import com.example.demo.dto.request.*;
import com.example.demo.dto.response.AuthResponse;
import com.example.demo.dto.response.UserResponse;
import com.example.demo.response.ApiResponse;
import com.example.demo.service.interfaces.AuthService;
import jakarta.validation.Valid;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.bind.annotation.*;

/**
 * Module 2 — Authentication. All endpoints are public.
 *
 * Access tokens only: there is no /refresh-token endpoint anywhere in this API.
 */
@RestController
@RequestMapping("/api/auth")
public class AuthController {
    public AuthController(AuthService authService) {
        this.authService = authService;
    }


    private final AuthService authService;

    /** POST /api/auth/register -> 201 Created */
    @PostMapping("/register")
    public ResponseEntity<ApiResponse<UserResponse>> register(
            @Valid @RequestBody RegisterRequest request) {

        UserResponse user = authService.register(request);
        return ResponseEntity.status(HttpStatus.CREATED).body(
                ApiResponse.success(
                        "Registration successful. Your membership number has been emailed to you.",
                        user));
    }

    /** POST /api/auth/login -> JWT access token */
    @PostMapping("/login")
    public ResponseEntity<ApiResponse<AuthResponse>> login(
            @Valid @RequestBody LoginRequest request) {

        return ResponseEntity.ok(ApiResponse.success("Login successful", authService.login(request)));
    }

    /**
     * Sign in with Google (OAuth 2.0 / OpenID Connect).
     *
     * The browser obtains an ID token from Google and posts it here; this
     * endpoint verifies it and returns OUR OWN JWT. From the next request
     * onwards a Google user is indistinguishable from a password user — same
     * token, same header, same authorization rules.
     *
     * Registers the user automatically on first sign-in, so there is no
     * separate "sign up with Google" endpoint to keep in step with this one.
     *
     * Public (permitAll) for the obvious reason: the caller has no token yet.
     */
    @PostMapping("/google")
    public ResponseEntity<ApiResponse<AuthResponse>> loginWithGoogle(
            @Valid @RequestBody GoogleLoginRequest request) {

        return ResponseEntity.ok(ApiResponse.success(
                "Signed in with Google", authService.loginWithGoogle(request)));
    }

    /**
     * POST /api/auth/logout
     *
     * With stateless JWT there is nothing to revoke on the server — the token
     * stays valid until it expires. The real logout is the frontend deleting
     * its copy. This endpoint exists so the client has something to call, and
     * it clears the SecurityContext for this request.
     *
     * (Server-side revocation would need a token blacklist table, which this
     * project deliberately does not have.)
     */
    @PostMapping("/logout")
    public ResponseEntity<ApiResponse<Void>> logout() {
        SecurityContextHolder.clearContext();
        return ResponseEntity.ok(ApiResponse.success(
                "Logged out. Please discard the access token on the client."));
    }

    /**
     * POST /api/auth/forgot-password
     *
     * Always returns 200, even for an unknown email — see AuthServiceImpl.
     */
    @PostMapping("/forgot-password")
    public ResponseEntity<ApiResponse<Void>> forgotPassword(
            @Valid @RequestBody ForgotPasswordRequest request) {

        authService.forgotPassword(request);
        return ResponseEntity.ok(ApiResponse.success(
                "If that email is registered, a password reset link has been sent to it."));
    }

    /** POST /api/auth/reset-password */
    @PostMapping("/reset-password")
    public ResponseEntity<ApiResponse<Void>> resetPassword(
            @Valid @RequestBody ResetPasswordRequest request) {

        authService.resetPassword(request);
        return ResponseEntity.ok(ApiResponse.success(
                "Password has been reset. You can now log in with your new password."));
    }
}

File: dotnet-backend/Emart.Api/Controllers/AuthController.cs
using System.Threading.Tasks;
using Emart.Api.DTOs;
using Emart.Api.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace Emart.Api.Controllers
{
    [ApiController]
    [Route("api/[controller]")]
    public class AuthController : ControllerBase
    {
        private readonly IAuthService _authService;

        public AuthController(IAuthService authService)
        {
            _authService = authService;
        }

        [HttpPost("register")]
        public async Task<ActionResult<ApiResponse<UserDto>>> Register([FromBody] RegisterRequest request)
        {
            var user = await _authService.RegisterAsync(request);
            return StatusCode(201, ApiResponse<UserDto>.CreateSuccess("Registration successful. Your membership number has been emailed to you.", user));
        }

        [HttpPost("login")]
        public async Task<ActionResult<ApiResponse<AuthResponse>>> Login([FromBody] LoginRequest request)
        {
            var response = await _authService.LoginAsync(request);
            return Ok(ApiResponse<AuthResponse>.CreateSuccess("Login successful", response));
        }

        [HttpPost("google")]
        public async Task<ActionResult<ApiResponse<AuthResponse>>> LoginWithGoogle([FromBody] GoogleLoginRequest request)
        {
            var response = await _authService.LoginWithGoogleAsync(request);
            return Ok(ApiResponse<AuthResponse>.CreateSuccess("Signed in with Google", response));
        }

        [HttpPost("logout")]
        [Authorize]
        public ActionResult<ApiResponse<object>> Logout()
        {
            // Stateless JWT logout is handled client-side.
            return Ok(ApiResponse<object>.CreateSuccess("Logged out. Please discard the access token on the client."));
        }

        [HttpPost("forgot-password")]
        public async Task<ActionResult<ApiResponse<object>>> ForgotPassword([FromBody] ForgotPasswordRequest request)
        {
            await _authService.ForgotPasswordAsync(request);
            return Ok(ApiResponse<object>.CreateSuccess("If that email is registered, a password reset link has been sent to it."));
        }

        [HttpPost("reset-password")]
        public async Task<ActionResult<ApiResponse<object>>> ResetPassword([FromBody] ResetPasswordRequest request)
        {
            await _authService.ResetPasswordAsync(request);
            return Ok(ApiResponse<object>.CreateSuccess("Password has been reset. You can now log in with your new password."));
        }
    }
}
